source/dub/packagemanager.d

```diff
@@ -1323,6 +1324,29 @@ symlink_exit:
 			}
 		}
 	}
+
+	/// Adds the package, scans for sub-packages, and returns the added package matching
+	/// the specified name (of the package itself or a sub-package).
+	/// Returns null if the sub-package doesn't exist.
+	private Package addPackagesAndResolveSubPackage(ref Package[] dst_repos, Package pack,
+		PackageName nameToResolve)
+	{
+		enforce(pack.name == nameToResolve.main.toString(),
+			"nameToResolve must be the added package or one of its sub-packages");
+
+		this.addPackages(dst_repos, pack);
+
+		if (nameToResolve.sub.length == 0)
+			return pack;
+
+		// available sub-packages have been appended
+		foreach_reverse (sp; dst_repos) {
+			if (sp.parentPackage is pack && sp.name == nameToResolve.toString())
+				return sp;
+		}
+
+		return null;
+	}
```

Comment:
I've totally reworked the fix, tackling the `PackageManager` API instead and resolving the subpackage in the lazy-loading `getPackage()` case (that was the regression AFAICT), as well as newly in `loadSCMPackage()`.

With more context now, I see that `resolveSubPackage()` is actually to be avoided, as it results in a full packages scan, killing the lazy-loading advantages. It's still used if the `dub.selections.json` contains a path-based selection, so currently we still get a full scan whenever depending on a subpackage of a path-selected parent package.

Reply:
Right, anything that calls `getPackageIterator` needs to go / be reworked.